Support stored procedures as transaction functions

dumbbell opened the pull request:
This patch brings two improvements:

1.  A **transaction function can take arguments**. Arguments are passed as a list of terms to `khepri:transaction()`.

    Here is an example:

    ```erlang
    generic_tx(Wood, Increase) ->
        Path = [stock, Wood],
        {ok, Stock} = khepri_tx:get(Path),
        NewStock = erlang:min(?MAX_STOCK, Stock + Increase),
        ok = khepri_tx:put(Path, NewStock).

    increase_wood_stock(Wood, Increase) ->
        khepri:transaction(
          ?STORE_ID, fun generic_tx/2, [Wood, Increase]).
    ```

    This way, one can write a more generic transaction function which doesn't rely on variables in its environment, and pass arguments to `khepri:transaction()` instead.

4.  **Stored procedures can be used as transaction functions**. This allows to extract and store a transaction function once and use it many times. This saves the extraction time and reduce the size of the transaction Ra command.

    Here is an example:

    ```erlang
    generic_tx(Wood, Increase) ->
        Path = [stock, Wood],
        {ok, Stock} = khepri_tx:get(Path),
        NewStock = erlang:min(?MAX_STOCK, Stock + Increase),
        ok = khepri_tx:put(Path, NewStock).

    store_wood_stock_tx() ->
        khepri:put(?STORE_ID, [tx, wood_stock], fun generic_tx/2).

    increase_wood_stock(Wood, Increase) ->
        khepri:transaction(
          ?STORE_ID, [tx, wood_stock], [Wood, Increase]).
    ```

    The constraints for transaction functions remain the same: they can't have side effects. So now, everytime a function is stored in Khepri, we verify during extraction if it can be used as a transaction function and store the result with the extracted function. When the stored procedure is executed as a transaction function, we check the result of that verification and possibly deny the execution.

Fixes #141.

the-mikedavis commented on src/khepri.erl:
```diff
@@ -2275,74 +2289,207 @@ transaction(Fun) ->
+%% the arity of `Fun'.
+%%
+%% Instead of `Fun', `PathPattern` can be passed. It must point to an existing
+%% stored procedure. The length ot `Args' must correspond to the arity of that
```
```suggestion
%% stored procedure. The length of `Args' must correspond to the arity of that
```

dumbbell replied:
Fixed.

the-mikedavis commented on src/khepri_payload.erl:
```diff
@@ -83,9 +84,15 @@ data(Term) ->
-    #p_sproc{sproc = Fun};
+    #p_sproc{sproc = Fun,
+             %% This will be overriden with the correct value when the
+             %% function will be extracted.
+             is_valid_as_tx_fun = false};
 sproc(#standalone_fun{} = Fun) ->
-    #p_sproc{sproc = Fun}.
+    #p_sproc{sproc = Fun,
+             %% This will be overriden with the correct value when the
```
there's a typo on these lines: overriden -> overridden

dumbbell replied:
Fixed.

the-mikedavis reviewed (commented):
Based on how I've seen transactions being used in the server, I think this will be a big usability improvement 🎉

dumbbell commented:
> Based on how I've seen transactions being used in the server, I think this will be a big usability improvement tada

Thank you!

I addressed your comments and force-pushed the changed.
